### PR TITLE
Support minicpm3-4b

### PR DESCRIPTION
**Pull request by AllentDan**
`update_weights` function is not obvious to users. @grimoire

### CODE REVIEW

**Comment by grimoire** [comment]
It is ok to copy weight into lm_head in load_weight (with the price that same weight would be doubled).

**Comment by lvhan028** [comment]
May update the supported models

**Comment by lvhan028** [comment]
@zhulinJulia24 Please add this model to test set

File: lmdeploy/pytorch/configurations/minicpm3.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from lmdeploy.pytorch.config import ModelConfig
+
+from .builder import AutoModelConfigBuilder
+
+
+class DeepseekV2ModelConfigBuilder(AutoModelConfigBuilder):
```
Comment (grimoire) [inline lmdeploy/pytorch/configurations/minicpm3.py:7]: Rename

File: lmdeploy/pytorch/models/minicpm3.py [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+                                    device=x.device,
+                                    dtype=x.dtype)
+
+        return (
+            self.cos_cached[:seq_len].to(dtype=x.dtype),
+            self.sin_cached[:seq_len].to(dtype=x.dtype),
+        )
+
+
+class MiniCPMLongRoPE(MiniCPMRotaryEmbedding):
```
Comment (grimoire) [inline lmdeploy/pytorch/models/minicpm3.py:89]: https://github.com/InternLM/lmdeploy/blob/5cedcb9a8742d65f94145d7ca2a0abcdc72be82f/lmdeploy/pytorch/backends/rotary_embedding.py#L15

**Comment by grimoire** [comment]
Can we use mla?

File: lmdeploy/pytorch/models/minicpm3.py (hunk 2 above)
Comment (AllentDan) [reply lmdeploy/pytorch/models/minicpm3.py:89]: Seems minicpm3-4b used a different rope scaling.

**Comment by AllentDan** [comment]
> Can we use mla?

I tried MLA and rope scaling like deepseekv2. But the result seemed wrong.

File: lmdeploy/pytorch/configurations/minicpm3.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                           num_layers=hf_config.num_hidden_layers,
+                           num_attention_heads=num_attention_heads,
+                           num_key_value_heads=num_key_value_heads,
+                           bos_token_id=hf_config.bos_token_id,
+                           eos_token_id=hf_config.eos_token_id,
+                           head_dim=head_dim,
+                           k_head_dim=k_head_dim,
+                           v_head_dim=v_head_dim,
+                           vocab_size=hf_config.vocab_size,
+                           multi_query_attention=True)
```
Comment (grimoire) [inline lmdeploy/pytorch/configurations/minicpm3.py:32]: Since you are not using mla, mqa should be false.

**Comment by lvhan028** [comment]
> May update the supported models
miss README

File: lmdeploy/pytorch/models/minicpm3.py [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., :x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2:]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
```
Comment (grimoire) [inline lmdeploy/pytorch/models/minicpm3.py:26]: We have provide apply rotary and rotary embedding op in `pytorch.nn`, Which is fused and would provide better performance.

**Comment by grimoire** [comment]
> I tried MLA and rope scaling like deepseekv2. But the result seemed wrong.

Less than 1000 blocks can be allocated for kv caches with `cache_max_entry_count=0.8` on A100. Add `TODO` comment if you don't want to optimize in this PR.

File: lmdeploy/pytorch/models/minicpm3.py [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+        quantization_config = getattr(config, 'quantization_config', None)
+        # gate up
+        self.gate_up_proj = build_merged_colwise_linear(
+            config.hidden_size,
+            [config.intermediate_size, config.intermediate_size],
+            bias=False,
+            dtype=dtype,
+            device=device,
+            quant_config=quantization_config,
+            is_tp=True,
```
Comment (lvhan028) [inline lmdeploy/pytorch/models/minicpm3.py:354]: If `tp=1` is passed to `PytorchEngineConfig`, do we still set `is_tp=True` here?

**Review by grimoire — APPROVED** [review]

**Review by irexyc — APPROVED** [review]
